feat(group-by-alerts): add group by alerts metric issue functionality

JoshFerge opened the pull request:
- Add functionality for Group By for the metric issue detector, specifically
  - handle grouped values returned from the snuba query subscription update data packet in `extract_value`
  - handle "missing" but active keys which may not be returned by subsequent snuba queries after initially triggering above threshold
  - modify `construct_title` and `create_occurrence` to take in a group key and use in title construction




Refs ID-836

[Tech Spec](https://www.notion.so/Group-By-Monitors-Tech-Spec-1b98b10e4b5d806faff8daee8bd0af43?source=copy_link)

JoshFerge committed:
feat(group-by-alerts): add group by alerts metric issue functionality

JoshFerge commented on tests/sentry/incidents/utils/test_metric_issue_base.py:
```diff
@@ -97,3 +102,130 @@ def process_packet_and_return_result(
+
+
+@freeze_time()
+class TestEvaluateMetricDetectorGroupBy(BaseMetricIssueTest):
```
would be helpful to walk through w/ someone on ACI team for best way to test

iamrajjoshi replied:
@mifu67 can probably help here

mifu67 replied:
Occurrence test looks good! I'd also suggest adding an integration test to make sure we go through post process and fire workflows correctly [in this file].(https://github.com/getsentry/sentry/blob/master/tests/sentry/incidents/test_metric_issue_post_process.py)

mifu67 replied:
I'd also want to verify that one group triggering and the other being below threshold works as expected.

JoshFerge commented on src/sentry/incidents/grouptype.py:
```diff
@@ -243,15 +249,42 @@ def extract_dedupe_value(self, data_packet: DataPacket[MetricUpdate]) -> int:
+        if "groups" in values:
+            # Return grouped data as dict[DetectorGroupKey, int]
+            grouped_values = {}
+            for group_data in values["groups"]:
```
this feels not ideal. would it be possible to type these data packet values somehow?

mifu67 replied:
`values` on the data packet's packet has `Any` type, so I don't see why not.

saponifi3d replied:
These are typed values, they're set when [defining the detector](https://github.com/getsentry/sentry/blob/master/src/sentry/workflow_engine/handlers/detector/base.py#L80) and then [enforced on `extract_value`](https://github.com/getsentry/sentry/blob/master/src/sentry/workflow_engine/handlers/detector/base.py#L130-L132)

Each of these will be typed and passed through as a `DataPacketEvaluationType`. i think it might be a little clearer on the typing etc if we were to set a flag like `has_group_by`. that would allow us to keep everything here pretty strongly typed.

mifu67 commented on tests/sentry/incidents/utils/test_metric_issue_base.py:
```diff
@@ -97,3 +102,126 @@ def process_packet_and_return_result(
+                resolution=timedelta(minutes=1),
+                environment=self.environment,
+                event_types=[SnubaQueryEventType.EventType.ERROR],
+                group_by=["level"],
```
If the detector trigger setup is the same, I'd recommend reusing the base class and updating the snuba query/self.detector_group_key with the group by info.

mifu67 replied:
For organization, we can add the issue occurrence test to [this file].(https://github.com/getsentry/sentry/blob/master/tests/sentry/incidents/test_metric_issue_detector_handler.py)

JoshFerge committed:
Merge branch 'master' into jferg/group-by-detector

JoshFerge committed:
Merge branch 'master' into jferg/group-by-detector

JoshFerge committed:
move test and add non-triggering group

JoshFerge committed:
fix types

JoshFerge committed:
fix anomoly detection

JoshFerge committed:
test(group-by-detector): Add integration tests for post-processing workflow

Add comprehensive integration tests for the group-by detector feature to verify
the full flow from grouped metric data through occurrence creation, post-processing,
and workflow execution.

Tests cover:
- Multiple groups triggering appropriate workflows with group keys in subtitles
- Priority escalation within specific groups
- Resolution workflow for grouped detectors

Addresses reviewer feedback requesting integration test coverage for group-by
detector functionality through post-process and workflow execution pipeline.

JoshFerge committed:
test(group-by-detector): Add test for missing group resolution behavior

Add test to verify behavior when a data packet contains no data for a group
that previously triggered an issue. This test currently fails, revealing that
missing group resolution behavior is not yet implemented.

The test serves as:
- Documentation of expected behavior (missing groups should resolve)
- Specification for future implementation
- Regression test once the behavior is implemented

Test expects that when a group is absent from subsequent data packets,
it should trigger a resolution status change for that group.

JoshFerge committed:
feat(group-by-detector): Implement missing group resolution behavior

When a group that previously triggered an alert is absent from new data packets,
automatically resolve that group while maintaining state for other groups.

- Add get_all_triggered_group_keys() to DetectorStateManager
- Add _get_missing_triggered_groups() helper to StatefulDetectorHandler
- Modify evaluate() to process missing groups and generate resolution messages
- Handle None condition_results for missing group scenarios

JoshFerge commented on src/sentry/workflow_engine/handlers/detector/stateful.py:
```diff
@@ -439,12 +451,59 @@ def evaluate(
                 data_value,
             )
 
+        # Process missing groups that were previously triggered but are now absent from data
```
let's move this into function

JoshFerge committed:
refactor: Extract missing groups processing into pure function

Refactor the missing group resolution logic to use a cleaner functional approach:
- _process_missing_triggered_groups() now returns results instead of mutating input
- Use early returns to reduce nesting and improve readability
- Better separation of concerns in the evaluate() method

JoshFerge committed:
Merge branch 'master' into jferg/group-by-detector

Commit:
clean up code

JoshFerge committed:
remove errant comment

saponifi3d commented on src/sentry/workflow_engine/handlers/detector/stateful.py:
```diff
@@ -439,12 +451,18 @@ def evaluate(
             )
 
+        # Process missing groups that were previously triggered but are now absent from data
+        missing_results = self._process_missing_triggered_groups(
+            group_data_values, data_packet, dedupe_value
+        )
```
might be nice to check if `group_data_values` is empty (like we don't have grouping) before invoking this. feel free to snag any code here: https://github.com/getsentry/sentry/pull/96172 i just needed to rename the flag 😓

saponifi3d commented on src/sentry/workflow_engine/handlers/detector/stateful.py:
```diff
@@ -298,6 +299,15 @@ def get_state_data(
 
+    def get_all_triggered_group_keys(self) -> list[DetectorGroupKey]:
+        """Get all group keys that are currently triggered for this detector"""
+        from sentry.workflow_engine.models.detector_state import DetectorState
```
DetectorState is already imported at the module level, could we just use that? might be a little more readable as `self.detector.detectorstate_set.filter(...)` as well, but i think that's a bit more pedantic.

saponifi3d commented on src/sentry/workflow_engine/handlers/detector/stateful.py:
```diff
@@ -366,9 +376,11 @@ def _build_workflow_engine_evidence_data(
-            "conditions": [
-                result.condition.get_snapshot() for result in evaluation_result.condition_results
-            ],
+            "conditions": (
+                [result.condition.get_snapshot() for result in evaluation_result.condition_results]
+                if evaluation_result is not None
+                else []
```
hmm, should we be invoking this method if we don't have an evaluation result? this is used to create the evidence to say _why_ a detector was triggered, if we don't have an evaluation result, then the detector is not triggered and we should be able to early exit evaluation for that group.

saponifi3d commented on src/sentry/incidents/grouptype.py:
```diff
@@ -243,15 +249,39 @@ def extract_dedupe_value(self, data_packet: DataPacket[MetricUpdate]) -> int:
         snuba_query: SnubaQuery,
         detector_trigger: DataCondition,
         priority: DetectorPriorityLevel,
+        group_key: DetectorGroupKey | None = None,
```
do we also need to update anything related to the fingerprints for groups? i think it'll be important that _each_ detector group key creates a unique fingerprint, so the issues are unique for users.

saponifi3d reviewed (commented):
overall really excited about this! lemme know if you have any questions or want to hop on a call to discuss anything, thanks for knocking this out!

JoshFerge committed:
add group by flag

JoshFerge committed:
use empty condition result instead of None

JoshFerge committed:
update mock detector state handler to have group by

saponifi3d commented on src/sentry/workflow_engine/handlers/detector/stateful.py:
```diff
@@ -298,6 +299,14 @@ def get_state_data(
             )
         return results
 
+    def get_all_triggered_group_keys(self) -> list[DetectorGroupKey]:
+        """Get all group keys that are currently triggered for this detector"""
+
+        triggered_states = self.detector.detectorstate_set.filter(is_triggered=True).exclude(
+            state=DetectorPriorityLevel.OK.value
+        )
+        return [state.detector_group_key for state in triggered_states]
```
🤔 i think we might want to have this work without state, so we can have the method live on the base detector. (since that's where the grouping info was added, it seems like we should support this w/o detector state)

saponifi3d commented on src/sentry/workflow_engine/handlers/detector/stateful.py:
```diff
@@ -567,6 +589,69 @@ def _get_configured_detector_levels(self) -> list[DetectorPriorityLevel]:
         conditions = self.detector.get_conditions()
         return list(DetectorPriorityLevel(condition.condition_result) for condition in conditions)
 
+    def _process_missing_triggered_groups(
```
it seems like this will resolve any group that isn't in the evaluation -- is that what we want to happen? i could see use cases where we just don't have data about that group, and we'd want to maintain the current state.

for example, for a grouping alert on p95 for /api/* to be 1s -- the first packet might look something like:
- `/api/projects/` p95 -> 1.2s -> Status => PriorityLevel.HIGH
- `/api/organizations/` p95 -> 0.3s -> Status => PriorityLevel.OK
The second packet is:
- `/api/organizations/` p95 -> 1.3s -> Status => PriorityLevel.HIGH

As a user, i'd expect both projects and orgs to still be in a high state, with this method it would resolve the issue for projects.

saponifi3d reviewed (commented):
i think my only question is around the processing_missing_triggered_groups and how we should handle that case. if i'm misunderstanding about how that data is coming back via snuba queries and we'll always have the group as long as it exists in the db, then we should be okay there too.
